### ISSUE TITLE
 Controller crash upon startup: libQt5Quick.so.5: undefined symbol

### ISSUE THREAD

**Issue by rischiopoco** [polychromatic/polychromatic#532]
### Details about the problem

When the program is launched, it always crashes giving the same error

### Terminal Output

```shell
$ polychromatic-controller -v
Traceback (most recent call last):
  File "/usr/bin/polychromatic-controller", line 34, in <module>
    from PyQt5.QtWebEngineWidgets import QWebEngineView
ImportError: /lib/x86_64-linux-gnu/libQt5Quick.so.5: undefined symbol: _ZN18QTextureGlyphCache8populateEP11QFontEngineiPKjPK11QFixedPoint, version Qt_5_PRIVATE_API
```

### Application Version

v0.9.3

### Qt Version

5.15.14

### Operating System

Ubuntu 22.04.5 LTS

### Desktop Environment

Xfce 4.16

**Comment by lah7** [comment]
I cannot replicate. The program works under Ubuntu 22.04 with the version (v0.8.7) it is packaged for:

```
$ polychromatic-controller  --version
Polychromatic 0.8.7
Python: 3.10.12
Qt: 5.15.6
```

Despite the versioning of the software, v0.8.x uses **Qt 5** and v0.9.x uses Qt 6, since some distros don't have PyQt6 packages (like Ubuntu 22.04). Both branches carry similar changes, so v0.8.7 and v0.9.3 are identical in terms of features.

Are you definitely trying to use **v0.9.3** under 22.04? Do you have some custom installation (pip?) going on?

    $ which python3
    /usr/bin/python3

    $ locate libQt5WebEngine.so
    /usr/lib/x86_64-linux-gnu/libQt5WebEngine.so.5
    /usr/lib/x86_64-linux-gnu/libQt5WebEngine.so.5.15
    /usr/lib/x86_64-linux-gnu/libQt5WebEngine.so.5.15.9

Usually, a crash like that with Qt libraries indicates some kind of interfering Python installation, libraries mismatching, using non-standard packages from a third party, or a partially updated system.

**Comment by rischiopoco** [comment]
Oh I'm sorry for the versioning, I may be wrong because I don't know how to check the version since I can't launch the controller.
I have just reinstalled Ubuntu 22.04.5 LTS doing a fresh installation and got this issue, it used to work fine.
I don't think I have custom installations, but how can I check if I have them?
Anyway, these are the command outputs:
```
$ which python3
/usr/bin/python3

locate libQt5WebEngine.so
(it doesn't give any output)
```
Sorry for the silly questions but I'm not very used to Linux systems

**Comment by lah7** [comment]
No worries! It is possible to check the version from the apt package system too:

    apt-cache show polychromatic-controller | grep Version

Anyway, since you got no output for `libQt5WebEngine.so`, that would indicate the dependency (`python3-pyqt5.qtwebengine`) doesn't seem to be installed, which is odd. Did you install Polychromatic using [our Ubuntu instructions](https://polychromatic.app/download/ubuntu)?

**Comment by rischiopoco** [comment]
Ok you were right, I have installed the 0.8.7 version
```
$ apt-cache show polychromatic-controller | grep Version
Version: 0.8.7
```
Yup, I followed the guide you linked, also I have tried to manually install the dependency `python3-pyqt5.qtwebengine` and I get this output message:
```
$ sudo apt-get install python3-pyqt5.qtwebengine
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
python3-pyqt5.qtwebengine is already the newest version (5.15.5-1).
python3-pyqt5.qtwebengine set to manually installed.
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
```
It says it is installed but I can't locate it

**Comment by lah7** [comment]
Let's try uninstalling the program (and its dependencies) in case it wasn't installed properly. If you happen to had reinstalled Ubuntu because things were acting up in the previous installation, beware if it's something like filesystem corruption or impending disk failure... _hopefully not._

Presumably, only our software depends on this web engine:

```
$ apt-cache rdepends --installed python3-pyqt5.qtwebengine
python3-pyqt5.qtwebengine
Reverse Depends:
  polychromatic-controller
  python3-pyqt5
```

and also:
```
$ apt-cache rdepends --installed libqt5webengine5
libqt5webengine5
Reverse Depends:
  libqt5webenginecore5
  python3-pyqt5.qtwebengine
  libqt5webenginecore5
```

I would uninstall the program, clean up and try again:

    sudo apt-get remove --purge polychromatic*
    sudo apt-get autoremove
    sudo apt-get autoclean

For good measure  (check this carefully in case you have other software that uses Qt 5 WebEngine):

    sudo apt remove libqt5webengine5 libqt5webenginecore5

Then install back and cross fingers:

    sudo apt install polychromatic

**Comment by rischiopoco** [comment]
I have reinstalled Ubuntu after updating the system to the new LTS version because I was experiencing a lot of problems (programs crash on startup as well as Xfce).
It seems like to be a common issue for users updating from LTS 22.04 to LTS 24.04 and didn't know that at the time.
After the update, I formatted the SSD and installed the old OS.
I have checked and got the same output messages for the first 2 commands that you wrote, so I proceeded to completely removed `polychromatic*` and `libqt5webengine5`, rebooted and reinstalled them.
Unfortunately, I'm still getting the same problem, but now the polychromatic symbol appears on the top right corner of the screen

**Comment by lah7** [comment]
Good to know, well... _not great_ on Ubuntu's side, but glad it's not the worst case scenario. This is a pretty uncommon and unlucky error.

That symbol in the top-right is likely the tray applet. That doesn't use QWebEngine - the web engine is used for the editor.

Looking again at the error, it fails on `libQt5Quick.so.5` which we don't use directly, but is needed for QWebEngine. By using a tool `plocate`, let's see if there's some stray version of that library on your system:

    sudo apt install plocate
    plocate libQt5Quick.so.5

It should output:

    /usr/lib/x86_64-linux-gnu/libQt5Quick.so.5
    /usr/lib/x86_64-linux-gnu/libQt5Quick.so.5.15
    /usr/lib/x86_64-linux-gnu/libQt5Quick.so.5.15.3

(When we uninstalled everything earlier, it should have uninstalled the package `libqt5quick5` too)

**Comment by rischiopoco** [comment]
Ok so the results are:
```
$ plocate libQt5Quick.so.5
(it doesn't give an output message)

$ md5sum /usr/lib/x86_64-linux-gnu/libQt5Quick.so.5*
6fccc69a89aee70198ce04127266ea73  /usr/lib/x86_64-linux-gnu/libQt5Quick.so.5
6fccc69a89aee70198ce04127266ea73  /usr/lib/x86_64-linux-gnu/libQt5Quick.so.5.15
6fccc69a89aee70198ce04127266ea73  /usr/lib/x86_64-linux-gnu/libQt5Quick.so.5.15.3
```
Btw thank you very much for the time you are dedicating to me, I really appreciate it :).
I don't know what to do to try to solve the problem, nothing like this has ever happened to me

**Comment by lah7** [comment]
No worries! Let's dig a little more:

When you installed `plocate`, it was supposed to update its index. Could try manually and see if `plocate` finds the library:

    sudo updatedb

Otherwise, is there something wrong with your paths?

    echo $PATH

It should roughly output:

    /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin

And nothing is overriding your library paths? This should output nothing:

    env | grep LD

As a fallback, if we can't locate where libQt5Quick.so.5 is, we could run:

    find / -name libQt5Quick.so.5 2>/dev/null

_Of course, the "nuclear" option is to just reinstall Ubuntu again a second time, and install OpenRazer + Polychromatic first to rule us out.. but it's probably not worth it_ 😁

**Comment by rischiopoco** [comment]
Oh yes, I had `plocate` already installed into the system and forgot to update its index, after updating it and running the command I got:
```
$ plocate libQt5Quick.so.5
/usr/lib/x86_64-linux-gnu/libQt5Quick.so.5
/usr/lib/x86_64-linux-gnu/libQt5Quick.so.5.15
/usr/lib/x86_64-linux-gnu/libQt5Quick.so.5.15.3
/usr/local/lib/python3.10/dist-packages/PyQt5/Qt5/lib/libQt5Quick.so.5
```
`echo $PATH` and `env | grep LD` give the expected output you wrote.

**Comment by lah7** [comment]
Found it! There's your culprit!

    /usr/local/lib/python3.10/dist-packages/PyQt5/Qt5/lib/libQt5Quick.so.5

I'm not sure what program this belongs to, but since it was installed outside the package system, the `/usr/local` folder overrides the `/usr` content, so it was causing a library mismatch. This could be caused if you ran `sudo pip install`  or `sudo make install` for something (which is discouraged for this reason).

You may wish to check what Python modules are in there, as well as any associated programs:

    ls /usr/local/lib/python3.10/dist-packages/
    ls /usr/local/bin

If you're confident these _local system-wide_ Python libraries are not important, you can clear the folder:

    sudo rm -rf /usr/local/lib/python3.10/dist-packages/*

If you did happen to run `sudo pip install`, a better option is to use [virtual environments](https://docs.python.org/3/library/venv.html), to avoid muddling up libraries with the system, like so:

    python -m venv venv
    source venv/bin/activate
    pip install <packages that won't interfere>

**Comment by rischiopoco** [comment]
Yes! It was indeed the problem, I deleted all `PyQt5` related files in `/usr/local/lib/python3.10/dist-packages/` and now it is working like a charm!
I will further investigate to find out what I have installed that needed that dependency, but at least now Polychromatic is working great!
Thank you so much for your time and for the patience to answer silly questions from a person who is not very experienced with Linux😁
